lib/Workflow.pm

```diff
@@ -315,6 +265,89 @@ sub set {
     $self->{$prop} = $value;
 }
 
+
+sub _execute_single_action {
+    my ( $self, $action_name, $autorun ) = @_;
+
+    # This checks the conditions behind the scenes, so there's no
+    # explicit 'check conditions' step here
+    my $action = $self->get_action($action_name);
+
+    # Need this in case we encounter an exception after we store the
+    # new state
+    my $old_state = $self->state;
+    my ( $new_state, $action_return );
+
+    eval {
+        $action->validate($self);
+        $self->log->is_debug && $self->log->debug("Action validated ok");
+        $action_return = $action->execute($self);
+        $self->log->is_debug && $self->log->debug("Action executed ok");
+
+        $new_state = $self->_get_next_state( $action_name, $action_return );
+        if ( $new_state ne NO_CHANGE_VALUE ) {
+            $self->log->is_info
+                && $self->log->info(
+                "Set new state '$new_state' after action executed");
+            $self->state($new_state);
+        }
+
+        # this will save the workflow histories as well as modify the
+        # state of the workflow history to reflect the NEW state of
+        # the workflow; if it fails we should have some means for the
+        # factory to rollback other transactions...
+
+        # Update
+        # Jim Brandt 4/16/2008: Implemented transactions for DBI persisters.
+        # Implementation still depends on each persister.
+
+        $self->_factory()->save_workflow($self);
+
+        # If using a DBI persister with no autocommit, commit here.
+        $self->_factory()->_commit_transaction($self);
```

Comment:
I've been wondering since this refactoring: shouldn't there be an observer message which the persister can hook into in order to commit its transaction?